README.md
<center>

# Limage

![Godot](https://img.shields.io/badge/Godot-3.4-blue?style=flat-square) ![GitHub](https://img.shields.io/github/license/teebarjunk/godot-limage?style=flat-square)  ![version](https://img.shields.io/badge/version-0.3-orange?style=flat-square)

Turn `.kra` `.psd` `.ora` images into *Godot* scenes `.tscn`.
</center>

Use your fav art tool. | What you see is what you'll get!
:-:|:-:
![layers1](./README/layers1.png) | ![layers3](./README/layers3.png)

Add origins + spawn points | to make life easier.
:-:|:-:
![points1](./README/points1.png) | ![points2](./README/points2.png)

Auto code generation, for a convenient inspector.

![layers2](./README/layers2.png)

Tell Godot what kind of node | you want the layer to be.
:-:|:-:
![node1](./README/node.png) | ![node2](./README/node2.png)

# Features
**Disclaimer:** Requires Python to be installed. Not all features work the same across all file formats, but I'm working on that.

![import_panel](./README/import_panel.png)

- Supports generating as `png` `webp` `jpeg` with tweakable settings for each.
- Scale options to shrink (or grow) all textures before importing.
- Optional [quantize](https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.quantize) to potentially shrink file size.
- [Origin](#Tags)s to make rotations easier.
- Main textures stored in the file, so it's all one asset.
- Secondary texture scenes are created for runtime load/unload.

# Getting Started
Requires python. In future I hope to reduce some of these steps.

- Download and install `limage` from [here](https://github.com/teebarjunk/limage).
- Drag `.kra` `.psd` `.ora` into Godot project.
- Select `import` in import tab.
- A scene will be generated next to the file.
- If any layers have the `options` or `toggles` tag, folders will be created with subscenes, that should stay next to the main scene.

# Node tags
Use `[]` for tags: `my layer [tag1 tag2]`\
Use `()` to apply tags to all children: `my group [tag] (childtag1 childtag2)`\
Use `(())` to apply tags to all descendants: `origins ((point))`

Tag|Description
:-:|--
`x`| Ignore layer. No texture or data generated.
`origin`| Defines origin of the parent group. If no parent, sets the global origin.
`node=?`| Hint for Godot to know what node to use, ie `[node=Light2D]`.
`group=?`| Add layer to Godot group. ie `[point group=Spawn]`
`point`| No texture. Center of image is converted to `Node2D` unless there is a `node` tag.
<!-- `copy`|(unfinished) Use texture of another layer. (Useful for limbs, eyes, repeating objects...)
`poly`|(unfinished)
`mask`|(unfinished) -->

## Group Layer Tags
Tag|Description
:-:|--
`options`|[[info]](#Options) Children will be options that can be toggled one at a time.
`toggles`|[[info]](#Options) Children will be toggleable.
`origins`|[[info]](#Origins) Children will be treated as points and used for layer origins, for easier rotations + scaling.
`points`|Children will all be turned into points, `Node2D`.
`merge`|"Flatten" child layers into one texture. (Not stable across formats.)

:o: = Tag for layer groups.\
Repository: [teebarjunk/limage](https://github.com/teebarjunk/limage)

# Options
When using `options` and `toggles`...

# Origins
todo

# Todo

- Better documentation.
- Points maintain layer order.
- Implement Krita clone layer/Photoshop smart layer.
- Finalize polygon generator.\*
- Auto normal map generation?
- [Sprite Dicing?](https://github.com/Elringus/SpriteDicing)
- Blend mode shaders?
- Vector layer support? (Looks doable for Krita.)

\* poly generation\
![polygons](./README/polygons.png)


# Solutions

## WEBP

If WEBP exporting isn't working, try installing, reinstalling, or updating [PILLOW](https://pillow.readthedocs.io/en/stable/installation.html), and/or libwebp:

On Ubuntu:

```
sudo apt-get install -y libwebp-dev
```

## Credits
<div>Script icon made by <a href="https://www.flaticon.com/authors/itim2101" title="itim2101">itim2101</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a></div>


CHANGES.md
# Changes

0.4
- Added export option `layers_as_scenes` saves a unique scene for each main layer.

0.3
- `kra` support added.
- Lot's of rewriting. Works more in line with Godot now.

0.2.3
- Added `padding` to settings.
- Added `copy` tag.
- Added `dir` tag.
- Added `keep_data` tag.
- Added setting `keep_data`.
- Added build toggles *force update* to force build to happen, and *skip images* to only update data.
- Made PNG the default image export format.
- Setting an origin sets for all children as well, making texture offsets easier.
- Origin names can include *-* to apply to a child.

0.2.2
- Added `node` tag.
- LimageNode now properly uses local coordinates, so it can be a child, rotated, scaled, without problems.
- LimageNode `options` and `toggles` fields now work in editor more stabily.
- LimageNode `toggles` initial visibility is fixed.

0.2.1
- Editor should reload files properly now.

0.2
- Saving + loading should work better on Windows.
- Fixed `merge` tag still generating child textures.
- Fixed JPEG error because of no alpha channel.
- Fixed polygon generation not working with some image formats.
- If settings file is changed, image will be rebuilt.
 


addons/limage/python/util.py
import sys, os, json

_print = print
indent = 0
def custom_print(*args):
	_print("\t" * indent + " ".join([str(x) for x in args]))

def get(data:dict, key:str, default:object = None):
	return default if key not in data else data[key]

def str_to_obj(s:str) -> object:
	try:
		return float(s)
	except:
		try:
			return int(s)
		except:
			l = s.lower()
			if l == "true": return True
			if l == "false": return False
			return s

def get_between(name:str, tag1="[", tag2="]"):
	s = name.find(tag1)
	e = name.find(tag2, s+len(tag1))
	data = {}
	if s != -1 and e != -1:
		inner = name[s+len(tag1):e]
		# print(f"INNER: [{inner}]")
		
		# Replace spaces between quotes, for a second
		while True:
			qs = inner.find('"')
			qe = inner.find('"', qs+1)
			if qs != -1 and qe != -1:
				q0 = inner[:qs]
				q = inner[qs+1:qe]
				qn = inner[qe+1:]
				inner = q0 + q.replace(" ", "####") + qn
			else:
				break
		
		name = name.replace(f"{tag1}{inner}{tag2}", "")#name[:s].strip() + name[e+len(tag2):]
		inner = inner.split(" ")
		for key in inner:
			if "=" in key:
				key, val = key.split("=")
				# Return spaces to quotes.
				val = val.replace("####", " ").strip()
				data[key] = str_to_obj(val)
			else:
				data[key] = True
	return name, data

def parse_name(name:str) -> tuple:
	name, descendants_data = get_between(name, "((", "))")
	name, child_data = get_between(name, "(", ")")
	name, data = get_between(name, "[", "]")
	
	# sanitize name
	output = ""
	for c in name:
		if c in "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789_ ":
			output += c
	output = output.strip()
	
	# make safe
	if False:
		name = name.lower().replace(" ", "_")
	
	# print(output, data, child_data, descendants_data)
	return output, data, child_data, descendants_data

def dict2str(data:dict) -> str:
	return json.dumps(data, separators=(',', ':'))

def file_size(path) -> str:
	return f"({nice_bytes(path.stat().st_size, 1)})"

def nice_bytes(size:int, precision:int=2) -> str:
	for suffix in ["b", "kb", "mb", "gb", "tp", "pb"]:
		if size > 1024:
			size /= 1024.0
		else:
			return f"{round(size, precision)} {suffix}"

addons/limage/python/godot.py
TEMPLATE_TRES:str = """[gd_resource type="Resource" load_steps=2 format=2]

[ext_resource path="res://addons/limage/Limage.gd" type="Script" id=1]

[resource]
script = ExtResource( 1 )
data = """

def get_options(layer) -> str:
	options = []
	default = ""
	for child in layer:
		options.append(child.name)
		if child._old_visible:
			default = child.name
	default = '"' + default + '"'
	options = '"' + '", "'.join(options) + '"'
	return options, default

def get_var_name(layer):
	return "_".join(layer._full_path).replace(" ", "_").replace("-", "_").lower()

def get_path(layer):
	return '"'+ "/".join(layer._full_path) +'"'

def generate_script(all_layers) -> str:
	has_script:bool = False
	head = "tool\nextends LimageNode\n\n"
	tail = ""
	for layer in all_layers:
		if "options" in layer._tags:
			has_script = True
			o, d = get_options(layer)
			v = get_var_name(layer)
			n = get_path(layer)
			head += f"export(String, {o}) var _{v}:String = {d} setget set_{v}, get_{v}\n"
			tail += f"func get_{v}() -> String: return get_option({n})\n"
			tail += f"func set_{v}(v:String):\n\t_{v} = v\n\tset_option({n}, v)\n\n" # \n\tif _{v} != v and v != \"\":\n\t\t_{v} = v\n\t\t
		
		if "toggles" in layer._tags:
			has_script = True
			for child in layer:
				v = get_var_name(child)
				d = "true" if child._old_visible else "false"
				n = get_path(child)
				head += f"export(bool) var _{v}:bool = {d} setget set_{v}, get_{v}\n"
				tail += f"func get_{v}() -> bool: return get_toggle({n})\n"
				tail += f"func set_{v}(v:bool):\n\t_{v} = v\n\tset_toggle({n}, v)\n\n" # \n\tif _{v} != v:\n\t\t # \n\t\t_{v} = v
	
	return has_script, f"{head}\n{tail}"

addons/limage/python/classes.py
class Vec2(dict):
	def __init__(self, x:float=0, y:float=0):
		if isinstance(x, (tuple, list)):
			self.x, self.y = x[0], x[1]
		elif isinstance(x, dict):
			self.x, self.y = x["x"], x["y"]
		elif isinstance(x, (float, int)):
			self.x, self.y = x, y
		else:
			print("VEC2 ERROR!")
	
	@property
	def x(self):
		return self["x"]
	
	@property
	def y(self):
		return self["y"]

	@x.setter
	def x(self, x):
		self["x"] = x
	
	@y.setter
	def y(self, y):
		self["y"] = y
	
	def __add__(self, obj):
		if isinstance(obj, (float, int)):
			return Vec2(self.x + obj, self.y + obj)
		elif isinstance(obj, (tuple, list)):
			return Vec2(self.x + obj[0], self.y + obj[1])
		elif isinstance(obj, dict):
			return Vec2(self.x + obj["x"], self.y + obj["y"])
		else:
			return Vec2(self.x + obj.x, self.y + obj.y)
	
	def __sub__(self, obj):
		if isinstance(obj, (float, int)):
			return Vec2(self.x - obj, self.y - obj)
		elif isinstance(obj, (tuple, list)):
			return Vec2(self.x - obj[0], self.y - obj[1])
		elif isinstance(obj, dict):
			return Vec2(self.x - obj["x"], self.y - obj["y"])
		else:
			return Vec2(self.x - obj.x, self.y - obj.y)
	
	def __mul__(self, obj):
		if isinstance(obj, (float, int)):
			return Vec2(self.x * obj, self.y * obj)
		elif isinstance(obj, (tuple, list)):
			return Vec2(self.x * obj[0], self.y * obj[1])
		elif isinstance(obj, dict):
			return Vec2(self.x * obj["x"], self.y * obj["y"])
		else:
			return Vec2(self.x * obj.x, self.y * obj.y)
	
	def negative(self):
		return Vec2(-self.x, -self.y)
	
	def copy(self):
		return Vec2(self.x, self.y)

addons/limage/python/genpoly.py
# NOT READY

# DONT USE

import numpy as np
import cv2 as cv

EXTRA_BORDER = 4
SIMPLIFY = 0.003

def get_points(pil_image, local_path) -> str:#path, local_path:str) -> list:
	# im = cv.imread(path)
	im = cv.cvtColor(np.array(pil_image), cv.COLOR_RGBA2BGRA)
	# cv.imwrite('./test_result0.png', im)
	
	_, mask = cv.threshold(im[:, :, 3], 0, 255, cv.THRESH_BINARY)
	
	# imgray = cv.cvtColor(im, cv.COLOR_BGRA2GRAY)
	
	# cv.imwrite('./test_result1.png', im)
	# cv.imwrite('./test_result2.png', imgray)
	
	# cv.imwrite('./test_result1.png', mask)
	
	_, thresh = cv.threshold(mask, 127, 255, cv.THRESH_BINARY)
	contours, hierarchy = cv.findContours(mask, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
	cv.drawContours(im, contours, -1, (0,0,0,255), EXTRA_BORDER)

	# cv.imwrite('./test_result2.png', im)
	
	_, mask = cv.threshold(im[:, :, 3], 0, 255, cv.THRESH_BINARY)
	
	# , cv.RETR_TREE   cv.RETR_EXTERNAL
	_, thresh = cv.threshold(mask, 127, 255, cv.THRESH_BINARY)
	contours, hierarchy = cv.findContours(thresh, cv.RETR_TREE, cv.CHAIN_APPROX_SIMPLE)
	
	if len(contours) == 0:
		print("no contours ", local_path)
		return None
	
	epsilon = SIMPLIFY * cv.arcLength(contours[0], True)
	approx = cv.approxPolyDP(contours[0], epsilon, True)
	cv.drawContours(im, [approx], -1, (0,255,0,255), 2)
	
	# cv.imwrite('./test_result3.png', im)
	
	points = ""
	uvs = ""
	for i in range(len(approx)):
		if i != 0:
			points += ", "
		x = approx[i][0][0]
		y = approx[i][0][1]
		points += f"{x}, {y}"
		uvs += f"{x}, {y}"
	
	return f"""[gd_scene load_steps=2 format=2]

[ext_resource path="{local_path}" type="Texture" id=1]

[node name="Polygon2D" type="Polygon2D"]
texture = ExtResource( 1 )
polygon = PoolVector2Array( {points} )
uv = PoolVector2Array( {uvs} )
"""
	
	# print(points)
	# print(len(contours[0]), len(approx))
	# cv.imwrite('./test_result.png', im)
	# cv.imshow('image',im)

addons/limage/python/process_kra.py
# NOT FUNCTIONAL, WIP

# DONT USE


import sys, zipfile, json
from PIL import Image
import xml.etree.ElementTree as ET
#import StringIO
HEAD = '{http://www.calligra.org/DTD/krita}'
def clean_head(child):
    return child.tag.replace(HEAD, "")

def recurse(d):
    key = clean_head(d)
    data = dict(d.attrib)
    data["__type"] = key
    children = []
    for child in d:
        children.append(recurse(child))
    if len(children) > 0:
        data["__children"] = children
    return data

    #sys.exit('Usage: '+sys.argv[0]+' <Input> <Output> <Size>')

zipf = zipfile.ZipFile("./block.kra")
maindoc = ET.ElementTree(ET.fromstring(zipf.read("maindoc.xml")))
#print(maindoc.getroot()[0].attrib)

data = recurse(maindoc.getroot(), 0)
print(json.dumps(data, indent=4))

#for child in maindoc.getroot():
    #key = clean_head(child)
    #print(key)
    #if key == "layers":
        #print(child[0].attrib["name"])
        #for layer in child[0][0]:
            #print("\t", clean_head(layer), layer.attrib['name'])

#.read('preview.png')

#im = Image.open(StringIO.StringIO(thumbnail))
#im.thumbnail( (int(sys.argv[3]), int(sys.argv[3])) )

#im.save(sys.argv[2],'png')



















"""
NOT USABLE AT ALL
probably never will be
"""

import glob
import os

import krita
from PyQt5 import QtWidgets


IMAGES_FILEPATH = QtWidgets.QFileDialog.getExistingDirectory()
FILEPATHS = sorted(glob.glob(IMAGES_FILEPATH + '/*'))
print(FILEPATHS)


def open_document(filepath, show=True):
	"""Open the given filepath as new document
	Arguments:
		filepath (str): the filepath of the image to open
		show (bool): should the image get shown in the Krita UI?
	"""
	
	k = krita.Krita.instance()
	print('Debug: opening %s' % filepath)
	doc = k.openDocument(filepath)
	if show:
		Application.activeWindow().addView(doc)
	return doc


def get_layers(doc):
	"""Return layers for given document"""
	nodes = []
	root = doc.rootNode()
	for node in root.childNodes():
		print('Debug: found node of type %s: %s' % (node.type(), node.name()))
		if node.type() == "paintlayer":
			nodes.append(node)
	return nodes



def make_layered_psd_from_images():
	"""Takes a folderpath, scans it for images and produces a layered image"""

	
	doc = open_document(FILEPATHS[0], show=False)
	doc_root = doc.rootNode()
	
	docs = []
	docs.append(doc)

	all_layers = get_layers(doc)
	for i in range(1, len(FILEPATHS)):
		docx = open_document(FILEPATHS[i], show=False)
		docs.append(docx)
		docx_layers = get_layers(docx)
		for layer in docx_layers:
			all_layers.append(layer.clone())
			# doc.rootNode().addChildNode(layer, parent_node)
	doc_root.setChildNodes(all_layers)

	print('Debug: all nodes: %s' % doc.rootNode().childNodes())
	# doc.refreshProjection()

	save_filepath = filepath = QtWidgets.QFileDialog.getSaveFileName()[0]
	r = doc.saveAs(save_filepath)
	print('Debug: saved: %s' % save_filepath)
	
	for doc in docs:
		print('Debug: closing %s' % doc)
		doc.close()

	print('Debug: Script done')


# make_layered_psd_from_images()






# # from krita import *
# # print(Krita.instance().filters())

# import zipfile
# import os
# #import PIL
# # from lxml import etree as ET
# import xml.etree.ElementTree as ET
# # tree = ET.ElementTree(ET.fromstring(xmlstring))


# # https://github.com/bloodywing/kraconvert/blob/master/kraconvert/kra.py

# def func():
# 	krafile = "./world_map.kra"
# 	kra = zipfile.ZipFile(krafile)

# 	filename = os.path.basename(krafile)
# 	basename, _ = filename.split('.')
	
# 	for k in kra.filelist:
# 		print(k)

# 	print(ET.fromstring(kra.read("maindoc.xml")["{http://www.calligra.org/DTD/krita}DOC"]))
# 	# merged_image = kra.read('mergedimage.png')

# 	# self.xml = ET.fromstring(kra.read('maindoc.xml'))
# 	# self.kra_name = self.xml.find('.//kra:IMAGE', ns).attrib['name']

# 	# self.icc = kra.read('{basename}/annotations/icc'.format(basename=self.kra_name))


# func()

addons/limage/python/PsdProcessor.py
import PIL
from PIL import Image
from PIL import features
from psd_tools import PSDImage
from psd_tools.constants import Tag
# from psd_tools.constants import BlendMode
import json, math, sys
from pathlib import Path
import util, godot
from util import get
from classes import Vec2

__version__ = "0.2"

WEBP_SUPPORTED:bool = features.check_module('webp')
if not WEBP_SUPPORTED:
	print(f"PILLOW v{PIL.__version__}")
	print(f"WEBP support: {WEBP_SUPPORTED}")
	print(f"  libwebp library might not be installed")
	print(f"  Ubuntu: sudo apt-get install -y libwebp-dev")

DEFAULT_SETTINGS:dict = {
	"path": "",						# location of psd
	"seperator": "-",				# change to "/" to folderize
	
	"texture_dir": None,			# if set, saves textures here
	"data_dir": "data",				# if set, saves layer data here
	# "script_dir": None,			# if set, saves godot scripts here
	
	# texture related
	"scale": 1,						# rescale textures
	"mask_scale": 1.0,#.25,			# shrink masks?
	"origin": [0, 0],				# multiplied by size of texture
	
	# can really decrease file size, but at cost of color range.
	# https://pillow.readthedocs.io/en/stable/reference/Image.html#PIL.Image.Image.quantize
	"quantize": False,				# decrease size + decrease quality
	
	# https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html
	# https://docs.godotengine.org/en/stable/getting_started/workflow/assets/importing_images.html
	"format": "WEBP",
	
	"PNG": {
		"optimize": True,
	},
	
	"WEBP": {
		"lossless": True,
		"method": 3,
		"quality": 80
	},

	"JPEG": {
		"optimize": True,
		"quality": 80
	}
}
FORCE_UPDATE:bool = False if len(sys.argv) <= 1 else sys.argv[1] == "True"
SKIP_IMAGES:bool = False if len(sys.argv) <= 2 else sys.argv[2] == "True"
DIR_RESOURCES = Path("/".join(sys.argv[0].split("/")[:-4]))
LOOK_IN:str = "layered_images"

def file_time(path) -> str:
	return str(path.lstat().st_mtime)

def file_exists(local_path) -> bool:
	path = DIR_RESOURCES / local_path
	return path.exists()

def create_directory(path):
	if not path.exists() and str(path).startswith(str(DIR_RESOURCES)):
		print(f"creating directory: {path}")
		path.mkdir(parents=True, exist_ok=True)

def load_dict(local_path, default:object=None) -> object:
	path = DIR_RESOURCES / local_path
	if path.exists():
		with open(path, "r") as file:
			return json.load(file)
	return default

def save_dict(local_path, data:dict, indent:bool=True, loud:bool=True):
	path = DIR_RESOURCES / local_path
	create_directory(path.parents[0])
	with open(path, "w") as file:
		if indent:
			json.dump(data, file, indent=4)
		else:
			json.dump(data, file, separators=(',', ':'))
	print(f"saved: {local_path}")

def save_string(local_path, string:str):
	path = DIR_RESOURCES / local_path
	create_directory(path.parents[0])
	with open(path, "w", encoding="utf-8") as file:
		file.write(string)
	print(f"saved: {local_path}")

def save_image(local_path, image, format, extension_settings):
	path = DIR_RESOURCES / local_path
	create_directory(path.parents[0])
	image.save(path, format, **extension_settings)
	print(f"saved: {local_path}")

class PSDProcessor:
	# settings
	def get(self, key:str, default=None):
		return get(self.settings, key, get(DEFAULT_SETTINGS, key, default))
	
	def __init__(self, path, settings:dict):
		self.settings = settings
		self.update_image = True
		self.path = path
		
		# filename without extension
		self.file_name = self.path.stem# self.path.rsplit("/", 1)[1].rsplit(".", 1)[0]
		self.name = self.get("name", self.file_name)
		self.texture_dir = self.get("texture_dir")
		if self.texture_dir == None:
			self.texture_dir = f"textures/{self.name}"
		
		self.data_dir = self.get("data_dir")
	
	def load(self):
		self.psd = PSDImage.open(self.path)
		self.wide, self.high = self.psd.size
		
		self.layers = []
		self.layer_info = {}
		self.texture_paths = []
		self.scenes = []
		
		self.output = {
			"name": self.name,
			"textures": self.texture_paths,
			"scenes": self.scenes
		}
		
		self.scale = self.get("scale")
		self.origin = Vec2(self.wide, self.high) * Vec2(self.get("origin"))
		self.mask_scale = self.scale * self.get("mask_scale")
		self.format = self.get("format")
		self.texture_extension = self.get("texture_extension", self.format.lower())
		self.extension_settings = dict(self.get(self.format, {}))
		self.structure = self.get("structure")
		
		
		# set default extension settings, if not set.
		if self.format in DEFAULT_SETTINGS:
			for key in DEFAULT_SETTINGS[self.format]:
				if not key in self.extension_settings:
					self.extension_settings[key] = DEFAULT_SETTINGS[self.format][key]
		
		layer_index = 0
		all_layers = list(self.psd.descendants())
		
		for l in list(all_layers):
			if l not in all_layers:
				continue
			
			# get name and tag data
			l.name, l._tags, l._child_tags, l._descendant_tags = util.parse_name(l.name)
			l.name = l.name if l.name != "" else f"LAYER_{layer_index}"
			l._index = layer_index
			layer_index += 1
			
			if "x" in l._tags:
				all_layers.remove(l)
				if l.kind == "group":
					for child in l.descendants():
						all_layers.remove(child)
				continue
			
			# remember old state
			l._old_visible = True if "visible" in l._tags else False if "!visible" in l._tags else l.visible
			l._old_opacity = l.opacity
			l._old_blend_mode = str(l.blend_mode).split(".")[1].lower()
			# force visible
			l.visible = True
			l.opacity = 255
			
			# get position + size
			x, y, r, b = l.bbox# self.get_bbox(l)
			x, y, r, b = max(x, 0), max(y, 0), min(r, self.wide), min(b, self.high)
			w = r - x
			h = b - y
			l._clamped_bbox = (x, y, r, b)
			l._size = Vec2(w, h)
			l._scale = self.scale
			l._origin = Vec2(x, y) + l._size * .5
			l._points = {}
			
		# update origins
		for l in list(all_layers):
			if l not in all_layers:
				continue
			
			if "point" in l._tags:
				pass
			
			if "origin" in l._tags:
				all_layers.remove(l)
				if l.parent == self.psd:
					self.origin = l._origin
				else:
					l.parent._origin = l._origin
			
			if "origins" in l._tags:
				all_layers.remove(l)
				for child_origin in l.descendants():
					all_layers.remove(child_origin)
					child = self.get_child(self.psd, child_origin.name)
					child._origin = child_origin._origin
					# print("set", child_origin.name, "origin to", child._origin)
		
		for l in list(all_layers):
			if l not in all_layers:
				continue
			
			# if merging, ignore lower data
			if "merge" in l._tags:
				for child in l.descendants():
					all_layers.remove(child)
			
			# determine if this is a group
			l._is_group = l.kind == "group" and not "merge" in l._tags
			if l._is_group:
				# pass on descendant data
				for child in l.descendants():
					for key in l._descendant_tags:
						if not key in child._tags:
							child._tags[key] = l._descendant_tags[key]
				
				# pass on child data
				for child in l:
					for key in l._child_tags:
						if not key in child._tags: 
							child._tags[key] = l._child_tags[key]
			
			del l._child_tags
			del l._descendant_tags
		
		for l in all_layers:
			l._path = self.get_layer_path(l)
			l._full_path = l._path.copy()
			l._full_path.append(l.name)
			
			if "scene" in l._tags:
				self.scenes.append(l)
			
			if not l._is_group and not "point" in l._tags:
				self.save_layer_image(l)
			
			# get rid of internally used keys
			for k in ["merge", "no_sep", "scene"]:
				if k in l._tags:
					del l._tags[k]
			
			l._global_position = Vec2(l._clamped_bbox[0], l._clamped_bbox[1])
			
		for l in all_layers:
			# l._origin -= l._global_position # localize
			
			# l._global_position += l._origin # offset origin
			# l._global_position -= self.origin
			
			l._origin *= l._scale
			l._global_position *= l._scale

			l._output = {
				"name": l.name,
				"path": l._full_path,
				"tags": l._tags,
				"visible": l._old_visible,
				"opacity": l._old_opacity,
				"blend_mode": l._old_blend_mode,
				"size": l._size * l._scale,
				"position": l._global_position,
				"origin": l._origin
			}
			
			if hasattr(l, "_texture"):
				l._output["scale"] = l._scale
				l._output["texture"] = l._texture
		
		self.all_layers = all_layers
		self.save_data()
	
	def was_modified(self, oldd:dict) -> bool:
		for key in self.settings:
			if key not in oldd:
				return False
		
		for key in oldd:
			if oldd[key] != self.settings[key]:
				return False
		
		return True
	
	def get_child(self, layer, child_name):
		for child in layer:
			if child.name == child_name:
				return child
		return None
	
	def save_data(self):
		if not self.data_dir:
			print("no 'data_dir' given. data won't be written to disk.")
			return
		
		for l in self.all_layers:
			if l._is_group:
				l._output["layers"] = [x._output for x in l if x in self.all_layers]
		
		# collect root layers
		output = {
			"name": self.file_name,
			"path": [],
			"tags": {"root": True},
			"visible": True,
			"opacity": 255,
			"size": Vec2(self.wide, self.high),
			"layers": [x._output for x in self.psd if x in self.all_layers],
			"blend_mode": "",
			"position": Vec2(0, 0),
			"origin": Vec2(0, 0),
		}
		
		def all_descendants(parent, layer, func):
			func(parent, layer)
			if "layers" in layer:
				for child in layer["layers"]:
					all_descendants(layer, child, func)
		
		# localize origin points
		def fix_origin(parent, child):
			if parent:
				child["origin"] -= child["position"]
				child["position"] += child["origin"]
		all_descendants(None, output, fix_origin)
		
		# localize layers
		def localize(parent, child):
			if parent:
				if not "options" in parent["tags"]: # options need to use parents origin.
					child["position"] -= parent["position"]
				child["position"] -= self.origin
		all_descendants(None, output, localize)
		
		# layer info
		tres_text = godot.TEMPLATE_TRES + json.dumps(output, indent=4)
		save_string(self.get_local_tres_path(), tres_text)
		
		# godot script
		has_script, script = godot.generate_script(self.all_layers)
		if has_script:
			script_name = self.name.replace(" ", "_")
			local_path = self.data_path(script_name, "gd")
			save_string(local_path, script)
	
	def get_local_tres_path(self):
		return self.data_path(self.name, "tres")
	
	def data_path(self, local_path, extension):
		return f"{self.data_dir}/{local_path}.{extension}"
	
	def get_layer_path(self, layer) -> list:
		path = []
		while layer.parent != None and layer.parent != self.psd:
			path.insert(0, layer.parent.name)
			layer = layer.parent
		return path
	
	def save_layer_image(self, l):
		if self.texture_dir == None:
			return
		
		local_path = self.texture_dir + "/"
		
		# add path to file name
		if not self.structure in ["flat", "solo"] and len(l._path) > 0:
			sep = self.get("seperator")
			local_path += sep.join(l._path) + sep
		
		local_path += l.name
		local_path += "." + self.texture_extension
		
		is_mask = "mask" in l._tags
		scale = self.scale if not is_mask else self.mask_scale
		
		l._scale = scale
		l._texture = f"res://{local_path}"
		self.texture_paths.append(local_path)
		
		if self.update_image and not SKIP_IMAGES:
			image = l.composite(l._clamped_bbox)
			
			# Scale.
			if scale != 1:
				w, h = image.size
				w = math.ceil(w * scale)
				h = math.ceil(h * scale)
				image = image.resize((w, h), Image.NEAREST if is_mask else Image.LANCZOS)
			
			# Optional: Quantize (Can really reduce size, but at cost of colors.)
			# 0 = median cut 1 = maximum coverage 2 = fast octree
			if is_mask:
				image = image.quantize(colors=2, method=2, dither=Image.NONE)
				
			elif self.get("quantize"):
				image = image.quantize(method=3)
			
			# generate polygon
			if "poly" in l._tags:
				import genpoly
				poly_path = self.data_path(f"poly_{self.name}", "tscn")
				points = genpoly.get_points(image, l._texture)
				save_string(poly_path, points)
			
			# RGBA -> RGB
			if self.format in ["JPEG"]:
				new_image = Image.new("RGB", image.size, (255, 255, 255))
				new_image.paste(image, mask=image.split()[3])
				image = new_image
			
			save_image(local_path, image, self.format, self.extension_settings)

def settings_changed(old, new):
	if len(old) != len(new):
		return True
	for k in new:
		if not k in old:
			return True
		if old[k] != new[k]:
			return True
	return False

directory = DIR_RESOURCES / LOOK_IN

if directory.exists():
	for file in list(directory.rglob('*.psd')):
		fname = file.stem
		fextension = file.suffix
		fpath = directory / file
		
		build_info_path = f"{LOOK_IN}/.lim_{fname}.json"
		build_info = load_dict(build_info_path, {"modified": "", "settings":{}})
		new_time = file_time(fpath)
		
		settings_path = f"{LOOK_IN}/{fname}.json"
		settings = load_dict(settings_path, {})
		
		psd_proc = PSDProcessor(fpath, settings)
		was_psd_modified = build_info["modified"] != new_time
		was_settings_modified = settings_changed(build_info["settings"], settings)
		was_data_deleted = not file_exists(psd_proc.get_local_tres_path())
		
		if FORCE_UPDATE or was_psd_modified or was_settings_modified or was_data_deleted:
			psd_proc.load()
			
			new_data = {
				"path": str(fpath),
				"modified": new_time,
				"settings": settings
			}
			save_dict(build_info_path, new_data)
		
		else:
			print(f"already up to date: {fname}")